cpp/src/io/json/reader_impl.cu

```diff
@@ -111,15 +72,169 @@ constexpr size_t calculate_max_row_size(int num_columns = 0) noexcept
+  const ParseOptions &opts,
+  cudaStream_t stream)
+{
+  // count keys
+  rmm::device_scalar<unsigned long long int> key_counter(0, stream);
+  auto const data_ptr = static_cast<const char *>(data.data());
+  cudf::io::json::gpu::collect_keys_info(data_ptr,
+                                         data.size(),
+                                         opts,
+                                         record_starts.data().get(),
+                                         record_starts.size(),
+                                         key_counter.data(),
+                                         nullptr,
+                                         stream);
+
+  // Allocate columns to store hash values, lengths, offsets
+  auto const num_keys = key_counter.value();
+  std::vector<std::unique_ptr<column>> info_columns;
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT64), num_keys));
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT16), num_keys));
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT32), num_keys));
+  auto info_table = std::make_unique<table>(std::move(info_columns));
+  rmm::device_scalar<mutable_table_device_view> info_table_mdv(
+    *mutable_table_device_view::create(info_table->mutable_view(), stream), stream);
+
+  // Reset the key counter - now used for indexing
+  key_counter.set_value(0, stream);
+  // Fill the allocated columns
+  cudf::io::json::gpu::collect_keys_info(data_ptr,
+                                         data.size(),
+                                         opts,
+                                         record_starts.data().get(),
+                                         record_starts.size(),
+                                         key_counter.data(),
+                                         info_table_mdv.data(),
+                                         stream);
+  return std::move(info_table);
+}
+
+/**
+ * @brief Extract the keys from the JSON file the name offsets/lengths.
+ */
+std::vector<std::string> create_key_strings(char const *h_data,
+                                            table_view sorted_info,
+                                            cudaStream_t stream)
+{
+  auto const num_cols = sorted_info.num_rows();
+  std::vector<uint64_t> h_offsets(num_cols);
+  cudaMemcpyAsync(h_offsets.data(),
+                  sorted_info.column(0).data<uint64_t>(),
+                  sizeof(uint64_t) * num_cols,
+                  cudaMemcpyDefault,
+                  stream);
+
+  std::vector<uint16_t> h_lens(num_cols);
+  cudaMemcpyAsync(h_lens.data(),
+                  sorted_info.column(1).data<uint16_t>(),
+                  sizeof(uint16_t) * num_cols,
+                  cudaMemcpyDefault,
+                  stream);
+
+  std::vector<std::string> names(num_cols);
+  std::transform(
+    h_offsets.begin(), h_offsets.end(), h_lens.begin(), names.begin(), [&](auto offset, auto len) {
```

Comment:
```suggestion
    h_offsets.cbegin(), h_offsets.cend(), h_lens.cbegin(), names.begin(), [&](auto offset, auto len) {
```

Reply:
@codereport how important is to use `cbegin()`? what are the benefits?
Plenty of places in our code simply uses `begin()`.

Reply:
AFAIK, it's the same as marking a variable as const. Mostly useful just to denote to the reader that the value will not change. Here, it would mean that we are not modifying `h_offsets`.

Reply:
Yes, it has many benefits:
1. It lets the reader know that you are not modifying the inputs. Many languages have syntax to denote modifying operations (aka the put ! at the end of those functions, such as Racket and other Lisps' **set!**). I mention this in my Algorithm Intuition talk - link to [minute 17:10](https://youtu.be/48gV1SNm3WA?t=1031) where I mention it
2. It will enable the compiler to tell you if you actually are modifying the inputs by accident
3. In some rare cases, it will enable the compiler to perform some more optimizations.